### PR TITLE
Implement new Codec that uses `mem.BufferSlice` instead of `[]byte`

### PR DESCRIPTION
**Pull request by PapaCharlie**
This new codec, as outlined in https://github.com/grpc/grpc-go/issues/6619 will allow the reuse of buffers to their fullest extent. Note that this deliberately does not (yet) implement support for stathandlers, however all the relevant APIs have been updated and both old and new Codec implementations are supported.

RELEASE NOTES:
- TBD

### CODE REVIEW

**Comment by easwars** [comment]
Could you please sign the CLA, and also fix the merge conflicts. Thanks.

**Comment by easwars** [comment]
FYI: @printchard

File: codec.go
Comment (easwars) [inline codec.go:80]: Following on from my other comment of having the encoder instance use the buffer pool configured on the channel, can we get that to happen here as well, so that all buffers come from the same pool?

File: internal/transport/controlbuf.go
Comment (easwars) [inline internal/transport/controlbuf.go:153]: One letter variable names in the existing code are not great. Can we name this `reader` instead?

File: internal/transport/http2_client.go
Comment (easwars) [inline internal/transport/http2_client.go:1204]: Similar here, we should get rid of the `nil` check.

And we can directly use the buffer pool from the transport here, instead of getting to the transport from the stream and then getting the buffer pool.

File: internal/transport/http2_server.go
Comment (easwars) [inline internal/transport/http2_server.go:819]: And here.

File: mem/buffers.go
Comment (easwars) [reply mem/buffers.go:218]: And if we can throw in a `Close` method on this `Reader` that frees up any unused buffers, we can make it implemented `io.ReadCloser` as well. Thoughts?

File: mem/buffers.go
Comment (easwars) [inline mem/buffers.go:286]: Why is this not a method on `BufferSlice` like `Reader()`?

**Comment by easwars** [comment]
@PapaCharlie : Did you get a chance to sign the CLA?

**Comment by easwars** [comment]
Would you mind splitting the `mem` package out to a separate PR?

We have some ongoing work that requires it, and it might be good to get that in before the rest to unblock that work. Thanks.

File: experimental/experimental.go
Comment (PapaCharlie) [reply experimental/experimental.go:54]: Oops

File: experimental/experimental.go
Comment (PapaCharlie) [reply experimental/experimental.go:37]: No, we definitely aren't. Using the buffer pool is expected to still be better, if not at least the same.

File: experimental/experimental.go
Comment (PapaCharlie) [reply experimental/experimental.go:45]: The buffer pool sadly _can't_ be used when those options are enabled, because the API is such that the binary loggers and tracing all expect the fully materialized buffers, instead of `mem.BufferSlice`.

File: experimental/shared_buffer_pool_test.go
Comment (PapaCharlie) [reply experimental/shared_buffer_pool_test.go:50]: I think my main concern with this TODO is that this test is now redundant. The entire test suite tests the buffer reuse now, so this test probably can just be deleted. I left myself a todo to validate that that was the case.

File: internal/transport/controlbuf.go
Comment (PapaCharlie) [reply internal/transport/controlbuf.go:977]: I think we talked about this. It's only supposed to be nil in tests, and I will add a comment to that effect.

Comment (PapaCharlie) [reply internal/transport/controlbuf.go:977]: Done

File: internal/transport/controlbuf.go
Comment (PapaCharlie) [reply internal/transport/controlbuf.go:502]: Sure, up to you

Comment (PapaCharlie) [reply internal/transport/controlbuf.go:502]: Done

File: encoding/proto/proto_v2.go
Comment (PapaCharlie) [reply encoding/proto/proto_v2.go:45]: I think we previously had a conversation on this. If the user really wants to have the specific pool that's used by the channel, they can provide a custom codec when initializing the client/server and make sure that that custom codec reuses the same `BufferPool` provided with `exeperimental.BufferPool`

File: encoding/proto/proto_v2.go
Comment (PapaCharlie) [reply encoding/proto/proto_v2.go:63]: See above

File: dialoptions.go
Comment (PapaCharlie) [reply dialoptions.go:761]: Oh good catch!

File: internal/transport/controlbuf.go
Comment (PapaCharlie) [reply internal/transport/controlbuf.go:153]: Can do!

File: internal/transport/http2_client.go
Comment (PapaCharlie) [reply internal/transport/http2_client.go:1204]: I added a comment on this, it will never be nil except for tests.

File: internal/transport/http2_server.go
Comment (PapaCharlie) [reply internal/transport/http2_server.go:819]: Same

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:29]: Cool makes sense

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:51]: Makes sense, added

File: mem/buffer_pool.go
Comment (PapaCharlie) [reply mem/buffer_pool.go:76]: Done!

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:98]: My reasoning for this one is that it's probably common to `defer buf.Free()` but potentially free the buffer even earlier, so subsequent frees should be a noop. It just seemed like good ergonomics, similar to how calling `delete` on a `map` doesn't mind if the key isn't present.

Comment (PapaCharlie) [reply mem/buffers.go:98]: Not super opinionated on this one though, happy to change it

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:151]: This doesn't actually print the slice, just the pointer to the first byte which helps uniquely identify the backing buffer. Printing the actual slice is actually _extremely_ noisy and I definitely do not recommend it! However I did like the suggestion of printing the length, which I added now

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:150]: Done

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:120]: Yep you are right. I have updated this

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:138]: Because the type of `free` is `func([]byte)`, meaning it needs the actual backing buffer to be passed in. Passing in `b.data[n:]` would not free the real buffer, only part of it

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:141]: This was buggy! I have fixed it now

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:117]: It'd mean copying over the panic message. Won't the compiler inline it for us?

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:39]: Fair enough

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:1]: ACK

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:62]: `func NewBufferWithCopy(data []byte)) *Buffer` is the same as `NewBuffer(slices.Clone(data), nil)` right? Doesn't seem like it adds too much value?

As for `Copy`, I really wanted it to be explicit that there's a copy happening. `NewBufferFromPool` doesn't express that as much. That being said, I appreciate it's also not super clear what it's actually doing, and I haven't been able to come up with a good name for it. If you still think `NewBufferFromPool` is a good name, I'll defer to you and change it

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:158]: 👍

Comment (PapaCharlie) [reply mem/buffers.go:158]: Formatted it like this:
```
// Len returns the sum of the length of all the Buffers in this slice.
//
// # Warning
//
// Invoking the built-in len on a BufferSlice will return the number of buffers
// in the slice, and *not* the value returned by this function.
```
This way there's a big section header even on gopkg. Thoughts?

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:161]: Sure, done

File: rpc_util.go
Comment (PapaCharlie) [reply rpc_util.go:622]: Shouldn't be necessary. We know it can't copy more than 5 bytes since `p.header` is an array, and `p.r.Read(5)` would have returned an error if the read was short.

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:207]: That's what makes it lazy: if all the data is already on one Buffer, it doesn't do anything. It'd be the opposite of `EagerMaterialize` I guess, where the buffer is _always_ copied, regardless of if it's all already on one buffer. Maybe `MaybeMaterialize`?

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:218]: ACK I can add that documentation. There's a specific reason for that, which is that `Reader` is used specifically in the `loopyWriter`, which can fail partially through the write and be retried later. This means freeing the buffers as the reader goes through them would make it impossible to retry. I can definitely add `Close` on it though, since it would remove needing to keep the reference to the original slice around in [`dataFrame`](https://github.com/PapaCharlie/grpc-go/blob/017b72fe45c6a1960468887642a0fc2f22771430/internal/transport/controlbuf.go#L152).

File: mem/buffers.go
Comment (PapaCharlie) [reply mem/buffers.go:286]: Because all the methods are on `(s BufferSlice)` and this would need to be on a pointer receiver (`func (s *BufferSlice) Writer(pool BufferPool) io.Writer`), which would be inconsistent and trigger linter warnings etc... If you feel strongly, I can just make it be a method on `BufferSlice`

File: rpc_util.go
Comment (tritone) [inline rpc_util.go:518]: Hey, I'm working on using the CodecV2 surface for a library that uses the CallOption https://pkg.go.dev/google.golang.org/grpc#ForceCodec currently to eliminate a data copy. Could you add a similar `ForceCodecV2` here so that we can customize a codec for the V2 API as well?

Comment (PapaCharlie) [reply rpc_util.go:518]: Oh whoops, I guess I missed it. I added the `ForceServerCodecV2` option but not the client one. Let me add that right now

Comment (tritone) [reply rpc_util.go:518]: Thank you!

Comment (PapaCharlie) [reply rpc_util.go:518]: Done! Have you been working with this PR? How has your experience been with it? That being said, what is the copy you are trying to get rid of specifically, and how are you doing it with the codec?

Comment (tritone) [reply rpc_util.go:518]: I'm working on using this for the Google Cloud Storage Go client (cloud.google.com/go/storage), which can use an experimental gRPC API in GCS. Specifically for object downloads, the data copy in protobuf unmarshaling was a pain point (because it copies the whole object data). See https://github.com/googleapis/google-cloud-go/pull/9526 for what we ended implementing to remove this.

I'm hoping this PR will help with the other major buffer alloc and copy that is still in the data path for downloads via gRPC. I'll be doing some more work on it this week as well as running CPU profiles, so I can let you know when I have more info from that :)

File: encoding/encoding_v2.go
Comment (PapaCharlie) [reply encoding/encoding_v2.go:3]: Just for the new files, presumably?

Comment (PapaCharlie) [reply encoding/encoding_v2.go:3]: Done

File: internal/transport/controlbuf.go
Comment (PapaCharlie) [reply internal/transport/controlbuf.go:995]: Chatted offline. When `hSize == 0`, the copy is a noop. Otherwise, the only way to get rid of this copy is to complete the framer changes

File: internal/transport/http2_client.go
Comment (PapaCharlie) [reply internal/transport/http2_client.go]: Yep, this actually doesn't compile otherwise anymore :)

File: internal/transport/http2_client.go
Comment (PapaCharlie) [reply internal/transport/http2_client.go:1216]: The function on the interface isn't needed anymore since the bufferPool is accessible through `t` directly, so I removed it. This code will also go away when the framer changes are added

File: internal/transport/http2_server.go
Comment (PapaCharlie) [reply internal/transport/http2_server.go]: Talked about this offline. It's less error prone to have `.Reader` take a reference and guarantee that it will be there rather than having to make sure `Reader` has its own reference. The overall attitude towards buffers is that callers to a function manage the lifecycle of buffers, and if a function wants to guarantee that a buffer lives beyond the function returning, it should always take its own reference. This has been updated in the codec as well

File: encoding/encoding_v2.go
Comment (dfawley) [inline encoding/encoding_v2.go:31]: Please document something like:  `The returned BufferSlice must have at least one reference that will be freed by gRPC when it is no longer needed.`

File: encoding/encoding_v2.go
Comment (dfawley) [inline encoding/encoding_v2.go:33]: Document e.g.: `data has a reference that is only valid for the duration of this call.  If the codec wishes, it may take its own reference to data that it Frees when it is no longer needed.`

**Review by dfawley — APPROVED** [review]

File: encoding/encoding_v2.go
Comment (ash2k) [inline encoding/encoding_v2.go:46]: Why not let users register codecs per-server and per-client? It's a more flexible approach and avoids error-prone globals e.g. no need for the `init()`. Global mutable state often leads to problems...

Comment (ash2k) [reply encoding/encoding_v2.go:46]: Some evil package somewhere in your import graph can register a `proto` codec that also sends messages to the homebase. Cannot happen with explicit registration requirements.

File: encoding/encoding_v2.go
Comment (PapaCharlie) [reply encoding/encoding_v2.go]: Update the docs

File: encoding/encoding_v2.go
Comment (PapaCharlie) [reply encoding/encoding_v2.go]: Ditto

File: encoding/encoding_v2.go
Comment (dfawley) [reply encoding/encoding_v2.go:46]: If something is in your binary, you are probably susceptible to even worse things than that.

I agree that globals are bad, but this is a pre-existing pattern that this PR isn't intended to address.

File: test/timeouts.go
Comment (easwars) [inline test/timeouts.go:25]: This needs to be reverted before merging.

**Review by easwars — APPROVED** [review]

File: test/timeouts.go
Comment (PapaCharlie) [reply test/timeouts.go:25]: Whoops!

**Review by dfawley — APPROVED** [review]

File: rpc_util.go
Comment (bboreham) [inline rpc_util.go:883]: Yes, it is massively more efficient.  See https://github.com/grpc/grpc-go/issues/2635, https://github.com/golang/go/issues/50774.

Comment (PapaCharlie) [reply rpc_util.go:883]: I have absolutely no doubt that in the context of buffers *not* being reused, it's more efficient. But in this case where buffers are pooled and produce as little garbage as possible, it may not? I don't know what the default buffer size used by the compressor is, but if it roughly lines up with one of the sized pools used by the `mem` package, then I'm not sure it would be more efficient to allocate one large buffer instead of representing the data in a `mem.BufferSlice`. That being said, it's 100% something that can be benchmarked and tested.

Comment (bboreham) [reply rpc_util.go:883]: Apologies, I had not caught up with that point.
A quick profile with gRPC 1.67 suggests `io.ReadAll` is no longer causing excess garbage, thanks!.

image

Comment (PapaCharlie) [reply rpc_util.go:883]: 🎉 🎉 🎉 

Awesome! That's really good to hear. Good to know this is a win for this new package. Do you think we would actually get a benefit from implementing this TODO? Or would you expect it to actually perform worse? If that's the case, maybe we can get rid of this commented block altogether instead

Comment (ash2k) [reply rpc_util.go:883]: Related to this function: https://github.com/grpc/grpc-go/pull/7653. `io.Copy()` makes lots of allocations - most for my use case.

Comment (bboreham) [reply rpc_util.go:883]: Great, I was hoping someone had the answer to that.

> Do you think we would actually get a benefit from implementing this TODO? 

I cannot think a major benefit right now.

**Comment by alanprot** [comment]
I think that this change should be more explicitly called out that the memory is being reused by default on the release notes as it can break some edge cases where the memory from the request is still being used even after the request ended.

The only think i can see on the release notes about this is:

```
codec: Implement a new Codec which uses buffer recycling for encoded message (https://github.com/grpc/grpc-go/pull/7356)
introduce a mem package to facilitate buffer reuse (https://github.com/grpc/grpc-go/pull/7432)
Special Thanks: @PapaCharlie
```

Which does not call out that the feature is enabled by default now.

**Pull request merged**
